File: 3-year/Concurrent Programming/teste2019/teste.java
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;

public class teste {
    public interface Jogo {
        Partida participa();
    }

    public interface Partida {
        String adivinha(int n);
    }

    Lock lock = new ReentrantLock();
    Condition cond = ;
    int numPlayers = 0;
    Partida curr = ;
    

    Partida participa() {
        lock.lock();
        try {
            numPlayers++;
            Partida myPartida = curr;
            if (numPlayers < 4) {
                while (myPartida == curr) {
                    cond.await();
                }
            } else {
                cond.signalAll();
                numPlayers = 0;
                curr = new Partida();
            }

            return myPartida;
        }
        finally {
            lock.unlock();
        }
    }

    public class Partida {
        int tries = 100;
        int num = 23; // num aleatorio entre 1 e 100

        long start = System.currentTimeMillis();

        Lock lock = new ReentrantLock(); 

        bool won = false;


        public String adivinha(int n) {
            lock.lock();
            try {
                if (won) {
                    return "PERDEU";
                }
                if (tries == 0) {
                    return "PERDEU";
                }
                if (System.currentTimeMillis() - start > 60000) {
                    return "PERDEU";
                }

                if (n == num) {
                    won = true;
                    return "GANHOU";
                } else if (n < num) {
                    tries--;
                    return "MAIOR";
                } else {
                    tries--;
                    return "MENOR";
                }

            } finally {
                lock.unlock();
            }
        }
    }
}

File: 3-year/Concurrent Programming/teste2019/teste.erl
% usamos processos e não threads
% não usamos partilha de memória

-module(teste).
-export([start/0, participa/0, jogo/1, createPartida/1, adivinha/2]).

start() ->
    Pid = spawn(fun() -> jogo([]) end),

    register(jogo_pid, Pid).

participa() ->
    jogo_pid ! {participa, self()} % enviamos uma mensagem ao jogo

    receive Msg -> Msg
    end.

jogo(Pids) ->
    receive
        {participa, Pid} -> % recebemos a mensagem do jogador
            NewPids = [Pid | Pids], % adicionamos o pid à lista de pids
            
            case length(Pids) of
                4 ->
                    createPartida(NewPids),
                    jogo([]); % reiniciamos o jogo
                _ ->
                    jogo(NewPids)
            end;
        end;
    end.

createPartida(Pids) ->
    P = spawn(fun() -> partida(100, rand:uniform(100)) end),
    [Pid ! P || Pid <- Pids], % enviamos o pid da partida a todos os jogadores
    spawn(fun() -> 
        receive
            after 60000 ->
                P ! partidaOver("TEMPO")
            end
        end,
    end),    
        
    )


adivinha(N, Partida) -> 
    Partida ! {N, self()},
    receive Msg -> Msg
    end.

partida(Tries, Num) ->
    receive 
        {N, Pid} -> % recebemos a mensagem do jogador
            if 
                N == Num ->
                    Pid ! "GANHOU",
                    partidaOver("PERDEU");
                N < Num ->
                    Pid ! "MAIOR",
                    partida(Tries - 1, Num);
                N > Num ->
                    Pid ! "MENOR",
                    partida(Tries - 1, Num)
            end;
        partidaOver ->
            partidaOver("TEMPO");
    end.

partidaOver(Status) ->
    receive
        {_, Pid} -> 
            Pid ! Status,
            partidaOver(Status);
        _ ->
            partidaOver(Status);
        partida(0, _) ->
            partidaOver("TENTATIVAS")
    end.
